### PR TITLE
Fix support for nullable types for AvroType attribute

### PR DESCRIPTION
**Pull request by robertcoltheart**
Fixes: #167 

@AdrianStrugala for your review please, this issue is preventing some key work on our side.

I'm slightly nervous that there are some edge cases with this or unepxected behaviours not catered for. For example, declaring a `byte[]` property, but using `AvroType(Decimal)` or something bizarre. I've added some additional unit tests to cover some of the edge cases I've found.

### CODE REVIEW

File: src/AvroConvert/AvroObjectServices/BuildSchema/ReflectionSchemaBuilder.cs [hunk 1]
```diff
@@ -479,18 +486,9 @@ public List<RecordFieldSchema> BuildRecordFields(
                             $"Type member '{info.MemberInfo.GetType().Name}' is not supported."));
                 }
 
-                TypeSchema fieldSchema;
-
-                if (info.AvroTypeRepresentation.HasValue)
```
Comment (robertcoltheart) [inline src/AvroConvert/AvroObjectServices/BuildSchema/ReflectionSchemaBuilder.cs:484]: This logic has been pushed down lower in the resolution tree, at the primitive/logical type level. This allows for unions to be created to support null types.

File: src/AvroConvert/AvroObjectServices/Write/Resolvers/Union.cs [hunk 2]
```diff
@@ -56,19 +57,19 @@ private bool UnionBranchMatches(TypeSchema typeSchema, object obj)
...
+                    return (obj is short or ushort or int or uint or char or byte or sbyte) || typeof(int).CanChangeTypeFrom(obj);
                 case AvroType.Long:
-                    return (obj is long or ulong);
+                    return (obj is long or ulong) || typeof(long).CanChangeTypeFrom(obj);
                 case AvroType.Float:
-                    return obj is float;
+                    return obj is float || typeof(float).CanChangeTypeFrom(obj);
                 case AvroType.Double:
-                    return obj is double;
+                    return obj is double || typeof(double).CanChangeTypeFrom(obj);
```
Comment (robertcoltheart) [inline src/AvroConvert/AvroObjectServices/Write/Resolvers/Union.cs:68]: These checks are needed for allowing conversions between declared property types, and actual values. Eg. a property type is an int but serialized/deserialized as a double (see unit tests).

**Comment by robertcoltheart** [comment]
@AdrianStrugala any chance of taking a look at this please?
